### PR TITLE
Port l1_loss to Aten

### PR DESCRIPTION
**Pull request by XiaobingSuper**
@VitalyFedyunin, This PR is about port L1 lose to Aten:
**Test script:**
```
import torch
import torch.nn as nn
import time

torch.manual_seed(0)

def _time():
    if torch.cuda.is_available():
        torch.cuda.synchronize()
    return time.time()

device = "cpu"
loss = nn.L1Loss(reduction = 'sum')
if torch.cuda.is_available():
    device = "cuda"
    loss = loss.cuda()

#warm up
for n in [100, 10000]:
    input = torch.randn(128, n, requires_grad=True, device=device)
    target = torch.randn(128, n, device=device)
    for i in range(1000):
        output = loss(input, target)
        output.backward()

#get running time
for n in [100, 10000]:
    fwd_t = 0
    bwd_t = 0
    input = torch.randn(128, n, requires_grad=True, device=device)
    target = torch.randn(128, n, device=device)
    for i in range(10000):
        t1 = _time()
        output = loss(input, target)
        t2 = _time()
        output.backward()
        t3 = _time()
        fwd_t = fwd_t + (t2 -t1)
        bwd_t = bwd_t + (t3 - t2)
    fwd_avg = fwd_t / 10000 * 1000
    bwd_avg = bwd_t / 10000 * 1000
    print("input size(128, %d) forward time is %.2f (ms); backwad avg time is %.2f (ms)."
          % (n, fwd_avg, bwd_avg))
```
Test Device: CPU: skx-8180, GPU: Tesla P100.

**Perfromance:**
Before:
```
GPU:
reduction=’mean’
nput size(128, 100) forward time is 0.31 (ms); backwad avg time is 0.09 (ms).
input size(128, 10000) forward time is 0.33 (ms); backwad avg time is 0.14 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.31 (ms); backwad avg time is 0.10 (ms).
input size(128, 10000) forward time is 0.34 (ms); backwad avg time is 0.14 (ms).

CPU:
reduction=’mean’
input size(128, 100) forward time is 0.06 (ms); backwad avg time is 0.10 (ms).
input size(128, 10000) forward time is 1.92 (ms); backwad avg time is 2.96 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.09 (ms).
input size(128, 10000) forward time is 1.96 (ms); backwad avg time is 2.79 (ms).

nume_thread = 1:
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 1.67 (ms); backwad avg time is 2.50 (ms).
reduction=’sum’:
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 1.67 (ms); backwad avg time is 2.51 (ms).
```
After:
```
GPU:
reduction=’mean’
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.10 (ms).
input size(128, 10000) forward time is 0.11 (ms); backwad avg time is 0.17 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.05 (ms); backwad avg time is 0.08 (ms).
input size(128, 10000) forward time is 0.11 (ms); backwad avg time is 0.16 (ms).

CPU:
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 0.14 (ms); backwad avg time is 0.18 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 0.15 (ms); backwad avg time is 0.17 (ms).

nume_thread = 1:
reduction=’mean’:
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.06 (ms).
input size(128, 10000) forward time is 1.05 (ms); backwad avg time is 1.72 (ms).
reduction=’sum’:
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 1.03 (ms); backwad avg time is 1.71 (ms).
```

How to set number thread? using following script:
```
num_threads=$1
script=$2
last_core=`expr $num_threads - 1`


echo "using $num_threads OMP threads"
echo "bind cores to 0~$last_core"

export OMP_NUM_THREADS=$num_threads
export KMP_AFFINITY=granularity=fine,compact,1,0

numactl --physcpubind=0-$last_core --membind=0 python $script
```
and run `./run.sh 1 L1loss.py`.

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
#24580 Migrate l1_loss from the TH to Aten (CUDA)
 #24581 Migrate l1_loss_backward from the TH to Aten (CUDA)
 #24717 Migrate l1_loss from the TH to Aten (CPU)
 #24718 Migrate l1_loss_backward from the TH to Aten (CPU)

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
 
     return apply_loss_reduction(loss, reduction);
 }
+
+Tensor l1_loss(const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto loss = input.sub(target).abs_();
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp]: This should cause performance regressions for smaller tensors. The original kernel in THNN was fused; now it is not fused. I think a better way to do this would be to port the entire function to use TensorIterator.

File: aten/src/ATen/native/Loss.cpp [hunk 2]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
 
     return apply_loss_reduction(loss, reduction);
 }
+
+Tensor l1_loss(const Tensor& input, const Tensor& target, int64_t reduction) {
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp]: Previously, the C++ implementation of l1_loss did not broadcast. Now, it does, so we can remove the `broadcast_tensors` line from
https://github.com/pytorch/pytorch/blob/8fb756d3b25be6b8f56bbd0d7b52b1f1ed1e7119/torch/nn/functional.py#L2177-L2180

File: aten/src/ATen/native/Loss.cpp [hunk 3]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
...
+  auto loss = input.sub(target).abs_();
+  return apply_loss_reduction(loss, reduction);
+}
+
+Tensor& l1_loss_out(Tensor&result, const Tensor& input, const Tensor& target, int64_t reduction) {
```
Comment (andreaskoepf) [inline aten/src/ATen/native/Loss.cpp]: Where is this function referenced/used?

File: aten/src/ATen/native/Loss.cpp [hunk 4]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
...
+    } else {
+      at::sum_out(result, loss, 0);
+    }          
+  } else {
+    at::sub_out(result, input, target).abs_();          
+  }
+  return result;
+}
+
+Tensor l1_loss_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction) {
```
Comment (andreaskoepf) [inline aten/src/ATen/native/Loss.cpp]: If a un-fused kernel is used, do we really need a custom backward functions or could we not better rely on auto_grad to do the job?

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/Loss.cpp]: @zou3519 I am not really sure if the extra complexity is worth this. The super-simple port with basic tensor operations already significantly improves the performance compared to the original fused implementation. At least for Tensors > 128x100. Before going the route of using tensor-iterator I would suggest to first perform further benchmarks with smaller tensors. e.g. see my test for soft_margin_loss  [here](https://github.com/pytorch/pytorch/pull/27673) which showed that a port even for 1000 element-tensors is faster.

File: aten/src/ATen/native/Loss.cpp (hunk 3 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/Loss.cpp]: ok, I see there still reamains an enty for `l1_loss.out` in `native_functions.yaml`. Are these special variants of the functions with destination tensor documented somewhere? The new/early entries in 'Loss.cpp' seem not have such special variants. I would prefer if all loss functions would share the same external interface and if the variant with provided destination tensor is important it should be documented somewhere.

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/Loss.cpp]: @andreaskoepf, there will have a big  performance degression if not using fused ops on GPU deveice for mse_loos which will call many other ops, please see https://github.com/pytorch/pytorch/pull/26529, for l1_loss, there only have a little perfromance degression  for the backward case, perhaps we don't need fused them, wait @VitalyFedyunin's response.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, Can you help see this PR. Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, code was rebased. Thanks!

**Comment by ezyang** [comment]
@XiaobingSuper this PR does not fix #27692. I wouldn't consider it blocking for landing here but I'm just making sure you know.

File: aten/src/ATen/native/Loss.cpp (hunk 3 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/Loss.cpp]: Please ignore my previous comment. The out-functions are required. The full set of forward, forward_out, backward & backward_out should be implemented.

**Comment by XiaobingSuper** [comment]
@ezyang, I add l1_loss test case, please help see it, thanks!

**Comment by ezyang** [comment]
Thanks! @VitalyFedyunin there are some comments here which are specifically waiting for you to weigh in, this PR has been open for a while and I'd like to move it along.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@ezyang, @VitalyFedyunin, can you help see this PR? Thanks!

File: aten/src/ATen/native/Loss.cpp (hunk 1 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/Loss.cpp]: I think we can live with tiny regression introduced by additional dispatch as we are planning to speed it up soon.

**Review by VitalyFedyunin — APPROVED** [review]

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Fails to merge, please rebase

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, rebased, thanks!
